clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

```diff
@@ -4466,8 +4476,133 @@ std::unique_ptr<mlir::Pass> createConvertCIRToLLVMPass() {
+    }
+    llvm_unreachable("unknown tbaa type");
+  }
+
+  mlir::ArrayAttr lowerTBAAToArrayAttr(mlir::MLIRContext *ctx,
+                                       std::optional<mlir::ArrayAttr> tbaa,
+                                       mlir::DataLayout &layout) {
+    if (!tbaa) {
+      return mlir::ArrayAttr();
+    }
+    auto tbaaTagArray = *tbaa;
+    if (tbaaTagArray.empty()) {
+      return mlir::ArrayAttr();
+    }
+    llvm::SmallVector<mlir::Attribute, 2> tags;
+    for (auto attr : tbaaTagArray) {
+      auto tag = lowerTBAAAttr(ctx, attr, layout);
+      tags.push_back(tag);
+    }
+    return mlir::ArrayAttr::get(ctx, tags);
+  }
+
+  mlir::LLVM::TBAATypeDescriptorAttr
+  createScalarTypeNode(mlir::MLIRContext *ctx, llvm::StringRef typeName,
+                       mlir::LLVM::TBAANodeAttr parent) const {
+    llvm::SmallVector<mlir::LLVM::TBAAMemberAttr, 2> members;
+    members.push_back(mlir::LLVM::TBAAMemberAttr::get(ctx, parent, 0));
+
+    return mlir::LLVM::TBAATypeDescriptorAttr::get(
+        ctx, typeName, llvm::ArrayRef<mlir::LLVM::TBAAMemberAttr>(members));
+  }
+
+  mlir::LLVM::TBAARootAttr getRoot(mlir::MLIRContext *ctx) const {
+    return mlir::LLVM::TBAARootAttr::get(
+        ctx, mlir::StringAttr::get(ctx, "Simple C/C++ TBAA"));
+  }
+
+  mlir::LLVM::TBAATypeDescriptorAttr getChar(mlir::MLIRContext *ctx) const {
+    return createScalarTypeNode(ctx, "omnipotent char", getRoot(ctx));
+  }
+
+  mlir::LLVM::TBAATypeDescriptorAttr
+  lowerScalarType(mlir::MLIRContext *ctx,
+                  cir::TBAAScalarTypeDescriptorAttr scalarType,
+                  mlir::DataLayout &layout) const {
+    // special handle for omnipotent char
+    if (auto intTy =
+            mlir::dyn_cast_or_null<cir::IntType>(scalarType.getType())) {
+      if (intTy.getWidth() == 1 || intTy.getWidth() == 8) {
+        return getChar(ctx);
+      }
+    }
+    auto name = getTypeName(scalarType.getType());
+    return createScalarTypeNode(ctx, name, getChar(ctx));
+  }
+
+  StringRef getTypeName(mlir::Type type) const {
+    return TypeSwitch<mlir::Type, StringRef>(type)
+        .Case<cir::IntType>([](cir::IntType ty) { return getIntTypeName(ty); })
+        .Case<cir::SingleType>([](cir::SingleType) { return "float"; })
+        .Case<cir::DoubleType>([](cir::DoubleType) { return "double"; })
+        .Case<cir::FP80Type>([](cir::FP80Type) { return "f80"; })
+        .Case<cir::FP128Type>([](cir::FP128Type) { return "f128"; })
+        .Case<cir::LongDoubleType>(
+            [](cir::LongDoubleType) { return "long double"; })
+        .Case<cir::BoolType>([](cir::BoolType) { return "bool"; })
+        .Case<cir::PointerType>([](cir::PointerType) { return "any pointer"; })
+        .Default([](auto ty) {
+          llvm::errs() << "unknown type: " << ty << "\n";
+          return "unknown";
+        });
+  }
+  static StringRef getIntTypeName(cir::IntType ty) {
+    switch (ty.getWidth()) {
```

Comment:
Such switch could be part of an extra declaration method in cir.IntType.

Reply:
added `getTBAATypeName()` in CIR_IntType's extraClassDeclaration.